File: im-desktop/src/services/forward-service.ts
import axios from 'axios';
import { ForwardRequest, ForwardResponse, BatchForwardResponse, MessageForward } from '../types/message-forward';
import { useAuthStore } from '../stores/auth-store';

const api = axios.create({
  baseURL: '/api/messages/forward',
  timeout: 15000,
});

api.interceptors.request.use((config) => {
  const token = useAuthStore.getState().token;
  if (token) {
    config.headers['X-User-Id'] = useAuthStore.getState().user?.id;
    config.headers['Authorization'] = `Bearer ${token}`;
  }
  return config;
});

export class ForwardService {
  static async forwardMessage(request: ForwardRequest): Promise<ForwardResponse> {
    const response = await api.post<ForwardResponse>('/', request);
    return response.data;
  }

  static async batchForward(requests: ForwardRequest[]): Promise<BatchForwardResponse> {
    const response = await api.post<BatchForwardResponse>('/batch', requests);
    return response.data;
  }

  static async getForwardHistory(messageId: number): Promise<MessageForward[]> {
    const response = await api.get<MessageForward[]>(`/history/${messageId}`);
    return response.data;
  }
}

export class ForwardHelper {
  static buildForwardContent(senderName: string, content: string, sentAt: string): string {
    return `[Forwarded]\nFrom: ${senderName}\nTime: ${sentAt}\n${'─'.repeat(15)}\n${content}`;
  }

  static buildMergedTitle(count: number): string {
    return `Merged ${count} messages`;
  }

  static async forwardToMultiple(
    messageIds: number[],
    targetIds: number[],
    merged = false,
    title?: string
  ): Promise<BatchForwardResponse> {
    const requests: ForwardRequest[] = targetIds.map(targetId => ({
      messageIds,
      targetConversationId: targetId,
      merged: merged && messageIds.length > 1,
      mergedTitle: title || (merged ? this.buildMergedTitle(messageIds.length) : undefined),
    }));
    return this.batchForward(requests);
  }
}

File: im-mobile/lib/services/forward_service.dart
import 'package:dio/dio.dart';
import '../models/message_forward.dart';
import '../models/conversation.dart';

class ForwardService {
  static final _dio = Dio(BaseOptions(
    baseUrl: '/api/messages/forward',
    connectTimeout: const Duration(seconds: 10),
    receiveTimeout: const Duration(seconds: 10),
  ));

  static Future<ForwardResponse> forwardMessage(ForwardRequest request) async {
    try {
      final response = await _dio.post('/', data: request.toJson());
      return ForwardResponse.fromJson(response.data);
    } on DioException catch (e) {
      return ForwardResponse(
        success: false,
        message: e.message ?? 'Network error',
        forwardedAt: DateTime.now(),
      );
    }
  }

  static Future<List<MessageForward>> getForwardHistory(int messageId) async {
    try {
      final response = await _dio.get('/history/$messageId');
      return (response.data as List)
          .map((e) => MessageForward.fromJson(e))
          .toList();
    } on DioException {
      return [];
    }
  }

  static String buildMergedTitle(int count) {
    return 'Merged $count messages';
  }

  static Future<List<ForwardResponse>> forwardToMultiple(
    List<int> messageIds,
    List<Conversation> targets, {
    bool merged = false,
    String? title,
  }) async {
    final futures = targets.map((t) async {
      return await forwardMessage(ForwardRequest(
        messageIds: messageIds,
        targetConversationId: t.id,
        merged: merged && messageIds.length > 1,
        mergedTitle: merged ? (title ?? buildMergedTitle(messageIds.length)) : null,
      ));
    }).toList();
    return await Future.wait(futures);
  }
}
